src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala

```diff
@@ -322,6 +326,27 @@ final class TrialService
     )
   }
 
+  private def finalizeUser(userInfo: UserInfo): Future[PerRequestMessage] = {
+    import TrialStates._
+
+    // get user's trial status, then check the current state
+    thurloeDao.getTrialStatus(userInfo.id, userInfo) flatMap { status =>
+      status.state match {
+        case Some(Finalized) =>
+          Future(RequestCompleteWithErrorReport(BadRequest, "Your free trial was already finalized."))
+        case Some(Terminated) =>
+          thurloeDao.saveTrialStatus(userInfo.id, userInfo, status.copy(state = Some(Finalized))) flatMap {
+            case Success(_) => Future(RequestComplete(NoContent))
+            case Failure(ex) => Future(RequestComplete(InternalServerError, ex.getMessage))
+          }
+        // user in some other state; can't finalize
+        case _ =>
+          val errMsg = "Your free trial should have been terminated first."
+          Future(RequestCompleteWithErrorReport(BadRequest, errMsg))
+      }
```

Comment:
it feels like this `match` duplicates the logic in `isAllowedFrom`. Could this block be:
```
if (Finalized.isAllowedFrom(status.state)) ... else ...
```
I'm also wondering if we should throw an error if the user is already Finalized. Consider the likely use case where a Terminated user has multiple tabs open, each of which displays the banner. If the user finalizes in one tab, then switches to the next and attempts to finalize again ... perhaps we should be lenient.
```
if (Finalized.isAllowedFrom(status.state)) {
  ... // persist Finalized state
} else {
  status.state match {
    case Some(Finalized) => Future(RequestComplete(NoContent)) // lenient noop
    case _ =>
      val errMsg = "Your free trial should have been terminated first."
      Future(RequestCompleteWithErrorReport(BadRequest, errMsg))
  }
}
```

Reply:
> perhaps we should be lenient

That's a good point. I agree that the system should be lenient. I'd personally make that happen in the UI since it's a user experience feature and to keep the backend consistent with the other status updates...but I don't feel strongly so went ahead and did it.